aten/src/ATen/Utils.h

```diff
@@ -83,9 +83,6 @@ static inline TensorImpl* checked_dense_tensor_unwrap(const Tensor& expr, const
     AT_ERROR("Expected object of scalar type ", scalar_type, " but got scalar type ", expr.scalar_type(),
              " for argument #", pos, " '", name, "' in call to ", api);
   }
-  if (expr.is_variable()) {  // TODO: change this to check `.requires_grad()` and `GradMode::is_enabled()` when Variable and Tensor are merged
-    AT_ERROR("Expected Tensor (not Variable) for argument #", pos, " '", name, "' in call to ", api);
-  }
```

Comment:
Is it correct that we will never enter TH when `VariableTensorId` is still in the tensor type set? I feel that if the dispatch mechanism gives us this guarantee, then we can remove the check here, otherwise checking `at::impl::variable_excluded_from_dispatch()` might be easier to reason about.

Reply:
TH legacy binding code always registers at a key like CPU or CUDA. You can't get there without handling Variable first. So there is indeed an invariant here; the error here is to guard against programmer error.